// ws/lib/SMError.js
/*
Copyright 2012 Microsoft Open Technologies, Inc.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

-----------------End of License---------*/


/**
 * Error codes
 */
module.exports = {
	/**
	 * Indicates normal closure.
	 */
	NORMAL: 1000,
	/**
	 * This is a generic error, and should only be used if a more specific error is not available.
	 */
	PROTOCOL_ERROR: 1002,
	/**
	 * This is returned when a frame is received for a stream which is not active.
	 */
	INVALID_STREAM: 2,
	/**
	 * Indicates that the stream was refused before any processing has been done on the stream.
	 */
	REFUSED_STREAM: 3,
	/**
	 * Indicates that the recipient of a stream does not support the SPDY version requested.
	 */
	UNSUPPORTED_VERSION: 4,
	/**
	 * Used by the creator of a stream to indicate that the stream is no longer needed.
	 */
	CANCEL: 5,
	/**
	 * This is a generic error which can be used when the implementation has internally failed, not due to anything in the protocol.
	 */
	INTERNAL_ERROR: 6,
	/**
	 * The endpoint detected that its peer violated the flow control protocol.
	 */
	FLOW_CONTROL_ERROR: 7,
	/**
	 * The endpoint received a SYN_REPLY for a stream already open.
	 */
	STREAM_IN_USE: 8,
	/**
	 * The endpoint received a data or SYN_REPLY frame for a stream which is half closed.
	 */
	STREAM_ALREADY_CLOSED: 9,
	/**
	 * The server received a request for a resource whose origin does not have valid credentials in the client certificate vector.
	 */
	INVALID_CREDENTIALS: 10,
	/**
	 * The endpoint received a frame which this implementation could not support.
	 */
	FRAME_TOO_LARGE:11
};
