// ws/lib/SMStream.js
/*
---------------------------------------
HTTPbis
Copyright Microsoft Open Technologies, Inc.
---------------------------------------
Microsoft Reference Source License.

This license governs use of the accompanying software. 
If you use the software, you accept this license. 
If you do not accept the license, do not use the software.

1. Definitions

The terms "reproduce," "reproduction," and "distribution" have the same meaning here 
as under U.S. copyright law.
"You" means the licensee of the software.
"Your company" means the company you worked for when you downloaded the software.
"Reference use" means use of the software within your company as a reference, in read only form, 
for the sole purposes of debugging your products, maintaining your products, 
or enhancing the interoperability of your products with the software, 
and specifically excludes the right to distribute the software outside of your company.
"Licensed patents" means any Licensor patent claims which read directly on the software 
as distributed by the Licensor under this license. 

2. Grant of Rights
(A) Copyright Grant- Subject to the terms of this license, the Licensor grants you a non-transferable, 
non-exclusive, worldwide, royalty-free copyright license to reproduce the software for reference use.
(B) Patent Grant- Subject to the terms of this license, the Licensor grants you a non-transferable, 
non-exclusive, worldwide, royalty-free patent license under licensed patents for reference use. 

3. Limitations
(A) No Trademark License- This license does not grant you any rights 
to use the Licensor’s name, logo, or trademarks.
(B) If you begin patent litigation against the Licensor over patents that you think may apply 
to the software (including a cross-claim or counterclaim in a lawsuit), your license 
to the software ends automatically. 
(C) The software is licensed "as-is." You bear the risk of using it. 
The Licensor gives no express warranties, guarantees or conditions. 
You may have additional consumer rights under your local laws 
which this license cannot change. To the extent permitted under your local laws, 
the Licensor excludes the implied warranties of merchantability, 
fitness for a particular purpose and non-infringement. 

-----------------End of License---------*/


var util = require('util'),
	events = require('events'),
	Error = require('./SMError');

/**
 * Speed+Mobility stream implementation
 */
function SMStream(streamId, session, headers, flags) {
	Object.defineProperty(this, 'streamId', { writable: false, value: streamId });
	Object.defineProperty(this, 'session', { writable: false, value: session });
	Object.defineProperty(this, 'state', { writable: true, value: SMStream.OPENED });
	Object.defineProperty(this, 'headers', { writable: false, value: headers || {} });
	
	var self = this;

	function onData(stream, data) {
		if (stream.streamId == self.streamId) {
			self.emit('data', data);
		}
	}

	function onHeaders(stream, headers) {
		if (stream.streamId == self.streamId) {
			self.emit('headers', headers);
		}
	}

	function onError(e) {
		if (e && e.streamId == self.streamId) {
			self.emit('error', e);
		}
	}

	session.on('data', onData);
	session.on('headers', onHeaders);
	session.on('error', onError);
	
	// stream is unidirectional
	if (flags && 0x02) self.state = SMStream.HALF_CLOSED;

	/**
	 * Closes stream
	 * @api public
	 */
	this.close = function(statusCode) {
		self.emit('close', statusCode);
	}

	/**
	 * Send data to the stream
	 * @api public
	 */
	this.send = function(data, fin) {
		if (self.state != SMStream.HALF_CLOSED && self.state != SMStream.CLOSED)
			self.emit('send', data, fin);
		else 
			self.error({statusCode:Error.STREAM_ALREADY_CLOSED, message:'Stream aready closed'});
	}

	/**
	 * Send headers to the stream
	 * @api public
	 */
	this.sendHeaders = function(headers, fin) {
		if (self.state != SMStream.HALF_CLOSED && self.state != SMStream.CLOSED)
			self.emit('sendHeaders', headers, fin);
		else 
			self.error({statusCode:Error.STREAM_ALREADY_CLOSED, message:'Stream already closed'});
	}

	/**
	 * Send error to the stream
	 * @api public
	 */	
	this.error = function(e) {
		self.emit('error', e);
	}
}

/**
 * Inherits from EventEmitter.
 */
util.inherits(SMStream, events.EventEmitter);

module.exports = SMStream;

/**
 * Stream States
 */
(function() {
  var StreamState = {
      OPENED: 0,
	  ACCEPTED: 1,
	  HALF_CLOSED: 2,
      CLOSED: 3
  };

  for (var state in StreamState) {
    if (!StreamState.hasOwnProperty(state)) continue;
    Object.defineProperty(SMStream, state, { enumerable: true, value: StreamState[state]});
  }
})();
