// ws/lib/SMStream.js
/*
Copyright 2012 Microsoft Open Technologies, Inc.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

-----------------End of License---------*/

var util = require('util'),
	events = require('events'),
	Error = require('./SMError');

/**
 * Speed+Mobility stream implementation
 */
function SMStream(streamId, session, headers, flags) {
	Object.defineProperty(this, 'streamId', { writable: false, value: streamId });
	Object.defineProperty(this, 'session', { writable: false, value: session });
	Object.defineProperty(this, 'state', { writable: true, value: SMStream.OPENED });
	Object.defineProperty(this, 'headers', { writable: false, value: headers || {} });
	
	var self = this;

	function onData(stream, data) {
		if (stream.streamId == self.streamId) {
			self.emit('data', data);
		}
	}

	function onHeaders(stream, headers) {
		if (stream.streamId == self.streamId) {
			self.emit('headers', headers);
		}
	}

	function onError(e) {
		if (e && e.streamId == self.streamId) {
			self.emit('error', e);
		}
	}

	session.on('data', onData);
	session.on('headers', onHeaders);
	session.on('error', onError);
	
	// stream is unidirectional
	if (flags == 0x02) self.state = SMStream.HALF_CLOSED;

	/**
	 * Closes stream
	 * @api public
	 */
	this.close = function(statusCode) { 
		self.emit('close', statusCode);
	}

	/**
	 * Send data to the stream
	 * @api public
	 */
	this.send = function(data, fin) {
		if (self.state != SMStream.HALF_CLOSED && self.state != SMStream.CLOSED) {
			self.emit('send', data, fin);
		} else {
			self.error({statusCode:Error.STREAM_ALREADY_CLOSED, message:'Stream already closed'});
		}
	}

	/**
	 * Send headers to the stream
	 * @api public
	 */
	this.sendHeaders = function(headers, fin) {
		if (self.state != SMStream.HALF_CLOSED && self.state != SMStream.CLOSED)
			self.emit('sendHeaders', headers, fin);
		else {
			self.error({statusCode:Error.STREAM_ALREADY_CLOSED, message:'Stream already closed'});
		}
	}

	/**
	 * Send error to the stream
	 * @api public
	 */	
	this.error = function(e) {
		self.emit('error', e);
	}
}

/**
 * Inherits from EventEmitter.
 */
util.inherits(SMStream, events.EventEmitter);

module.exports = SMStream;

/**
 * Stream States
 */
(function() {
  var StreamState = {
      OPENED: 0,
	  ACCEPTED: 1,
	  HALF_CLOSED: 2,
      CLOSED: 3
  };

  for (var state in StreamState) {
    if (!StreamState.hasOwnProperty(state)) continue;
    Object.defineProperty(SMStream, state, { enumerable: true, value: StreamState[state]});
  }
})();
